### ISSUE TITLE
Refactor to use go/analysis

### ISSUE THREAD

**Issue by alexandear** [alexkohler/nakedret#24]
I propose refactoring `nakedret` to use [`go/analysis`](https://pkg.go.dev/golang.org/x/tools/go/analysis). `go/analysis` is an API for modular analysis. Also, it’s a common interface for all linters.

 After refactoring will be possible:
- to run `nakedret` with `go vet -vettool=$(which nakedret)`;
- and to properly integrate with `golangci-lint` (see https://github.com/golangci/golangci-lint/pull/3760).

For implementation details see [nakedret.go](https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/nakedret.go) and [blog post](https://disaev.me/p/writing-useful-go-analysis-linter/).

**Comment by zalenskivolt** [comment]
@alexandear when using
```go
func (in *Inspector) Preorder(types []ast.Node, f func(ast.Node)) {
```

is it possible to keep track of when the nodes pop? In this case, the code looks for `*ast.FuncDecl`, `*ast.FuncLit`, and `*ast.ReturnStmt`, but it needs to keep track of which function each return belongs to, so when an inner function ends and there is another return statement in the outer function, it needs to know that the parsing is back in the outer one.

**Comment by zalenskivolt** [comment]
> using
> 
> ```go
> func (in *Inspector) Preorder(types []ast.Node, f func(ast.Node)) {
> ```
> 
> is it possible to keep track of when the nodes pop?

it should use one of
```go
func (in *Inspector) Nodes(types []ast.Node, f func(n ast.Node, push bool) (proceed bool))
func (in *Inspector) WithStack(types []ast.Node, f func(n ast.Node, push bool, stack []ast.Node) (proceed bool)) {
```

**Comment by alexandear** [comment]
@zalenskivolt sorry, but I'm not the author of nakedret and can't answer your questions.

**Comment by zalenskivolt** [comment]
@alexkohler I did a take at using go/analysis in #25, also adding suggested fixes as in #8 using that framework.

This replaces main with the provided `go/analysis/singlechecker.Main()`, and that should also mean that all the file handling in checkNakedReturns, parseInput, etc could be removed. For now I left the old tests in there, but the new `analysistest.Run` and `analysistest.RunWithSuggestedFixes` tests should be able to test everything.

I updated the copy to match what's currently in the `golangci-lint` version.

I would love to get this going, to get an updated version properly integrated in `golangci-lint`.

**Comment by alexkohler** [comment]
@zalenskivolt - thanks for the PR! Took a look and would be happy to merge. 

> I updated the copy to match what's currently in the golangci-lint version.

Just to clarify: does `golangci-lint` currently  have a copy of `nakedret` or the like that's been refactored to use `go/analysis`,  and your PR just ports those changes upstream?

**Comment by zalenskivolt** [comment]
> @zalenskivolt - thanks for the PR! Took a look and would be happy to merge.

Great! 😄  A small detail when actually testing this version with golangci-lint: The "main" application needs to be in a different place than any code you want to import from outside. I think the smallest change is to add a `cmd/nakedret/main.go` with just the `main()` function. I'll push that too.

> Just to clarify: does `golangci-lint` currently have a copy of `nakedret` or the like that's been refactored to use `go/analysis`, and your PR just ports those changes upstream?

Yes, there's a copy in [golangci-lint/pkg/golinters/nakedret.go](https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/nakedret.go), a small test in [test/testdata/nakedret.go](https://github.com/golangci/golangci-lint/blob/master/test/testdata/nakedret.go), and [third_party/nakedret/LICENSE](https://github.com/golangci/golangci-lint/blob/master/third_party/nakedret/LICENSE).
It just wraps an earlier version of the `ast.Walk` code with a tiny go/analysis wrap and some golangci-lint-specific issue reporting. This PR uses only go/analysis, and I think that's the case for many linters. I copied the error message from the golangci-lint version.

I tried updating the golangci-lint copy to the most recent version in https://github.com/golangci/golangci-lint/pull/3760, but got some push-back:

> the way that we handle this linter is not right, we have to change that.
> 
> Instead of "duplicating" the code inside golangci-lint I think it's better to a create PR on nakedret to be able to use it as a dependency.

hence this PR 😄

**Comment by zalenskivolt** [comment]
pushed `main()` &rarr; `cmd/nakedret/main.go`

**Comment by alexkohler** [comment]
Merged #25 and tagged this with [v1.0.2](https://github.com/alexkohler/nakedret/releases/tag/v1.0.2), please let me know if you need anything else!
